### PR TITLE
Downloading and using Sentinel-1 Estonia data

### LINKED ISSUE #43: Downloading and using Sentinel-1 Estonia data
Once the Sentinel-1 Estonia data is uploaded to Zenodo, the following features need to be implemented:
- Download of S1 data from Zenodo. User should be able to decide whether to download S1 and/or S2 data
- Option to choose which data to use for training: S1 and/or S2

### PR DESCRIPTION
**Pull request by jsreuss**
Closes #43

### CODE REVIEW

File: eurocropsml/dataset/utils.py [hunk 1]
```diff
@@ -26,29 +26,32 @@
...
 ) -> tuple[dict[int, list[str]], dict[int, list[str]] | None]:
-    files_list = read_files(data_dir)
+    files_list: set = set()
+    for s in satellite:
+        files_list.update(read_files(data_dir.joinpath(s)))
 
     # pre-filter by regions for region and regionclass split
     if regions is not None:
-        files_list = [file for file in files_list if file.startswith(tuple(regions))]
+        files_list = {file for file in files_list if file.startswith(tuple(regions))}
```
Comment (jmaces) [inline eurocropsml/dataset/utils.py:54]: Maybe change the variable naming to reflect that this is not a "list" but a set?

File: eurocropsml/dataset/utils.py [hunk 2]
```diff
@@ -60,7 +63,7 @@ def _split_dataset(
 
 
 def _filter_classes(
-    file_list: list[str],
+    file_list: set,
```
Comment (jmaces) [inline eurocropsml/dataset/utils.py:66]: Also here, the argument name contradicts the type annotation. Also can we make the annotation more specific? A set of what do we expect here? (Previously it was strings, is there a reason why this was removed?)

File: eurocropsml/dataset/utils.py (hunk 1 above)
Comment (jsreuss) [reply eurocropsml/dataset/utils.py]: Also done

File: eurocropsml/dataset/dataset.py [hunk 3]
```diff
@@ -48,107 +48,161 @@ def __init__(
...
-            array_name: torch.tensor(np_array) for array_name, np_array in arrays_dict.items()
+        meta_data: dict[str, dict[str, torch.Tensor] | torch.Tensor] = {
+            array_name: {key: torch.tensor(np_array) for key, np_array in meta_array.items()}
+            for array_name, meta_array in arrays_dict.items()
         }
+        # center is always the same, replace by first value
+        meta_data["center"] = cast(torch.Tensor, next(iter(meta_data["center"].values())))
 
-        tensor_data = torch.tensor(np_data, dtype=torch.float)
         if self.config.normalize:
```
Comment (jsreuss) [inline eurocropsml/dataset/dataset.py:153]: @jmaces: I am thinking of removing the if-clause and just always apply the normalization. Otherwise, in case no normalization is applied, all our padding values etc do not fit anymore.

File: eurocropsml/dataset/dataset.py [hunk 4]
```diff
@@ -48,107 +49,164 @@ def __init__(
...
-        tensor_data = torch.tensor(np_data, dtype=torch.float)
+        # normalization and scaling to (0,1]
         if self.config.normalize:
-            tensor_data = torch.mul(tensor_data, NORMALIZING_FACTOR)
+            if "S1" in np_data_dict:
+                normalized_s1 = np_data_dict["S1"] * NORMALIZING_FACTOR_S1
+                np_data_dict["S1"] = (normalized_s1 + 1) / 2 + EPSILON
+            if "S2" in np_data_dict:
+                normalized_s2 = np_data_dict["S2"] * NORMALIZING_FACTOR_S2
+                np_data_dict["S2"] = normalized_s2 * (1 - EPSILON) + EPSILON
```
Comment (jsreuss) [inline eurocropsml/dataset/dataset.py:161]: @jmaces: I am thinking of removing the if-clause and just always apply the normalization. Otherwise, in case no normalization is applied, all our padding values etc. do not fit anymore.

File: eurocropsml/dataset/dataset.py [hunk 5]
```diff
@@ -48,107 +49,164 @@ def __init__(
...
-            array_name: torch.tensor(np_array) for array_name, np_array in arrays_dict.items()
+        meta_data: dict[str, dict[str, torch.Tensor] | torch.Tensor] = {
+            array_name: {key: torch.tensor(np_array) for key, np_array in meta_array.items()}
+            for array_name, meta_array in arrays_dict.items()
         }
+        # center is always the same, replace by first value
+        meta_data["center"] = cast(torch.Tensor, next(iter(meta_data["center"].values())))
 
-        tensor_data = torch.tensor(np_data, dtype=torch.float)
+        # normalization and scaling to (0,1]
```
Comment (jsreuss) [inline eurocropsml/dataset/dataset.py:154]: @jmaces: I am thinking of removing the if-clause and just always apply the normalization. Otherwise, in case no normalization is applied, all our padding values etc. do not fit anymore.

File: eurocropsml/dataset/dataset.py (hunk 4 above)
Comment (jsreuss) [inline eurocropsml/dataset/dataset.py:161]: @jmaces: I am thinking of removing the if-clause and just always apply the normalization. Otherwise, in case no normalization is applied, all our padding values etc. do not fit anymore.

File: eurocropsml/dataset/dataset.py [hunk 6]
```diff
@@ -71,88 +78,135 @@ def __init__(
...
         if self.config.normalize:
-            tensor_data = torch.mul(tensor_data, NORMALIZING_FACTOR)
+            if "S1" in np_data_dict:
+                normalized_s1 = np_data_dict["S1"] * NORMALIZING_FACTOR_S1
+                np_data_dict["S1"] = (normalized_s1 + 1) / 2 + EPSILON
+            if "S2" in np_data_dict:
+                normalized_s2 = np_data_dict["S2"] * NORMALIZING_FACTOR_S2
+                np_data_dict["S2"] = normalized_s2 * (1 - EPSILON) + EPSILON
```
Comment (jmaces) [inline eurocropsml/dataset/dataset.py:166]: Could we add a comment for both if-cases what the expected range of values before normalization is? That would help to clarify how the goal to scale them to (0,1] is achieved.

File: eurocropsml/dataset/dataset.py (hunk 4 above)
Comment (jmaces) [reply eurocropsml/dataset/dataset.py:166]: I agree that now it makes no sense to use our default padding without normalization. but I guess the padding could also be modified by, e.g., using a different custom `collate_fn`. Maybe instead of removing the option to deactivate normalization we could instead add a warning that this would require custom modification of the padding as well?

File: eurocropsml/dataset/dataset.py [hunk 7]
```diff
@@ -71,88 +78,135 @@ def __init__(
...
 
         if self.pad_seq_to_366:
-            tensor_data = pad_seq_to_366(tensor_data, meta_data["dates"])
+            for satellite, value_array in np_data_dict.items():
+                np_data_dict[satellite] = pad_seq_to_366(
+                    value_array, meta_data["dates"][satellite]  # type: ignore[index]
+                )
+            meta_data["dates"] = _unique_dates(
+                cast(dict[str, torch.Tensor], meta_data["dates"]), list(f.keys())
```
Comment (jmaces) [inline eurocropsml/dataset/dataset.py:174]: Minor: If we use Iterable instead of List for the last `_unique_dates` argument would could avoid converting to list here (and in a few more places as well). (Not a big concern though, so we could leave as it is, as these lists will always be quite short)

File: tests/dataset/test_dataset.py [hunk 8]
```diff
@@ -20,80 +16,75 @@ def preprocess_config() -> EuroCropsDatasetPreprocessConfig:
...
         arrays_dict=leap_test_arrays_dict,
-        config=config,
         preprocess_config=preprocess_config,
         s2_data_bands=S2_BANDS,
     )
-    data = arrays_dict["data"]
-    dates = arrays_dict["dates"]
-    assert data.shape == leap_test_arrays_dict["data"].shape
+    data = arrays_dict["data"]["S1"]
+    dates = arrays_dict["dates"]["S1"]
```
Comment (jmaces) [inline tests/dataset/test_dataset.py:69]: Just to double check: Do we only have tests for "S1" now?

File: tests/dataset/test_dataset.py [hunk 9]
```diff
@@ -20,80 +16,75 @@ def preprocess_config() -> EuroCropsDatasetPreprocessConfig:
...
         arrays_dict=test_arrays_dict,
-        config=config,
         preprocess_config=preprocess_config,
         s2_data_bands=S2_BANDS,
     )
-    data = arrays_dict["data"]
-    dates = arrays_dict["dates"]
-    assert data.shape == (12, test_arrays_dict["data"].shape[1])
+    data = arrays_dict["data"]["S1"]
+    dates = arrays_dict["dates"]["S1"]
```
Comment (jmaces) [inline tests/dataset/test_dataset.py:87]: Same question as above.

File: tests/dataset/test_dataset.py (hunk 8 above)
Comment (jsreuss) [reply tests/dataset/test_dataset.py:69]: Good point! I will loop through both satellites.

File: eurocropsml/dataset/dataset.py (hunk 6 above)
Comment (jsreuss) [reply eurocropsml/dataset/dataset.py]: Done :)

File: tests/dataset/test_dataset.py (hunk 9 above)
Comment (jsreuss) [reply tests/dataset/test_dataset.py]: See above

File: eurocropsml/dataset/dataset.py [hunk 10]
```diff
@@ -159,9 +159,11 @@ def __getitem__(
         # normalization and scaling to (0,1]
         if self.config.normalize:
             if "S1" in np_data_dict:
+                # range before normalization and scaling: [-50.0,1.0]
                 normalized_s1 = np_data_dict["S1"] * NORMALIZING_FACTOR_S1
                 np_data_dict["S1"] = (normalized_s1 + 1) / 2 + EPSILON
```
Comment (jmaces) [inline eurocropsml/dataset/dataset.py]: In this case, we would not use the full target range (0,1] for S1, but only (0, 0.51+eps] no? I would expect the scaling to be
```python
normalized_s1 = (np_data_dict["S1"]+50) * NORMALIZING_FACTOR_S1
np_data_dict["S1"] = normalized_s1 * (1-EPSILON) + EPSILON
```
with `NORMALIZING_FACTOR_S1=1/51`

Comment (jsreuss) [reply eurocropsml/dataset/dataset.py]: Ah yes, you're right!

**Review by jmaces — DISMISSED** [review]

**Review by jmaces — DISMISSED** [review]

File: eurocropsml/dataset/download.py [hunk 11]
```diff
@@ -96,10 +96,10 @@ def select_version(versions: list[dict]) -> tuple[dict, list[str]]:
         if selected_id >= 8:
 
             logger.warning(
-                "Please be aware that for Zenodo version 8 or older and this package version "
+                "Please be aware that Zenodo version 8 or older and this package version "
                 "(eurocropsml>=0.4.0) are not compatible anymore in terms of "
                 "eurocropsml.preprocess.preprocess. The already peprocessed version of Sentinel-2 "
```
Comment (jmaces) [inline eurocropsml/dataset/download.py:101]:
```suggestion
                "eurocropsml.preprocess.preprocess. The already preprocessed version of Sentinel-2 "
```

File: eurocropsml/dataset/download.py [hunk 12]
```diff
@@ -96,10 +96,10 @@ def select_version(versions: list[dict]) -> tuple[dict, list[str]]:
...
-                "can still be used, but re-running the preprocessng will will not filter out all "
+                "can still be used, but re-running the preprocessing will will not filter out all "
```
Comment (jmaces) [inline eurocropsml/dataset/download.py:102]:
```suggestion
                "can still be used, but re-running the preprocessing will not filter out all "
```

**Review by jmaces — APPROVED** [review]
